### PR TITLE
Michael/weapon 2

### PR DESCRIPTION
**Pull request by michaelwiest**
This lays out the core functionality for having weapons that can have multiple distinct attacks. The weapon also denotes how long the user has between attacks to continue the "attack string" before it getting reset to the first attack.
The general structure is:
```
Weapon
  - attack_i
    - animation
    - hitbox
  - attack_j
```
This is mainly due to the fact that for some reason inherited classes (eg a weapon in this case) can't alter an animationPlayer. So each attack has a single animationPlayer that has a `use` function.

There are a few helper functions for when an attack is going etc. None of those are SUPER helpful yet until we can connect the weapon to the player via an inventory. 

IGNORE THE BELOW I FIGURED IT OUT.

The other annoying thing is that i couldn't add attack nodes directly to an exported var from the Weapon class. That is directly add the appropriate attack nodes to the scene tree. Instead we use an exported var which is an array of attack scene names. We then dynamically add and connect the attacks to the Weapon. It's fine but annoying.

### CODE REVIEW

**Comment by michaelwiest** [comment]
Lotta changes in here but the main changes are to Weapon and Attack. I cleaned up some stuff in Hitbox

File: game/scripts/core/Attack.gd [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@onready var animation_player = $AnimationPlayer
+@onready var hitbox = $Marker2D/hitbox
+@onready var sprite = $Sprite2D
+	
+func use():
+	animation_player.play("use")
+
+func attack_start():
+	enable()
+	emit_signal("attack_start_signal")
```
Comment (zshainsky) [inline game/scripts/core/Attack.gd]: Small suggestion I think the preferred syntax is 

`attack_start_signal.emit()`

File: game/scripts/core/Attack.gd [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+func disable():
+	hitbox.disable()
+	sprite.visible = false
+
+func attack_end():
+	disable()
+	emit_signal("attack_end_signal")
+
+# Kind of hacky signaling to pass a hit to the parent weapon. Because the
+# hitbox has no notion of the force or damage that it can apply.
```
Comment (zshainsky) [inline game/scripts/core/Attack.gd]: This seams like a totally legit way to do this

File: game/scripts/core/Weapon.gd [hunk 3]
```diff
@@ -1,67 +1,70 @@
...
 
-# Called every frame. 'delta' is the elapsed time since the previous frame.
-func _process(delta):
-	pass
-#
-#func use():
-#	UseAnimation.play("use")
-#
 
 func equip(equipment_slot):
```
Comment (zshainsky) [inline game/scripts/core/Weapon.gd:41]: Wonder if we will need this here. We can discuss how to connect these componentes inventory, player, weapon

File: game/scripts/inventory/weapon_data.gd [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+enum EquipmentSlot {
+	LIGHT,
+	HEAVY,
+	PROJECTILE
+}
+
+@export var force: float
+@export var damage: float
+@export var stops_movment: bool 
+@export var equipment_slot: EquipmentSlot
```
Comment (zshainsky) [inline game/scripts/inventory/weapon_data.gd:13]: I was thinking having an EquipmentSlot makes sense so probably will want to add this logic to the inventory. Maybe a new class that extends SlotData. Something to think about

**Review by zshainsky — APPROVED** [review]
This is legit. Got it working on my machine today. I think the way you designed it makes a lot of sense!

**Pull request merged**